src/Plugin/SmsGateway/Telstra2.php

```diff
@@ -0,0 +1,358 @@
+      }
+      else {
+        $result->setError(SmsMessageResultStatus::ERROR);
+        $result->setErrorMessage((string) $this->t("Response code '@code'", [
+            '@code' => $status,
+        ]));
+      }
+
+      return $result;
+    }
+
+    $messages = $response->getMessages();
+    $message_id = reset($messages)->getMessageId();
+
+    $report
+      ->setRecipient($recipient)
+      ->setMessageId($message_id)
+      ->setStatus(SmsMessageReportStatus::QUEUED)
+      ->setTimeQueued(REQUEST_TIME);
+    $result->setReports([$report]);
+
+    return $result;
+  }
+
+  /**
+   * Process an incoming message POST request.
+   *
+   * Telstra only passes one message per request.
+   *
+   * API documentation:
+   *   https://dev.telstra.com/content/sms-getting-started#callback
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request object.
+   * @param \Drupal\sms\Entity\SmsGatewayInterface $sms_gateway
+   *   The gateway instance.
+   *
+   * @return \Drupal\sms\SmsProcessingResponse
+   *   A SMS processing response task.
+   */
+  function processIncoming(Request $request, SmsGatewayInterface $sms_gateway) {
+    $messages = [];
+
+    $json = Json::decode($request->getContent());
+
+    $time_atom = $json['acknowledgedTimestamp'];
+    $time = DrupalDateTime::createFromFormat(DATE_ATOM, $time_atom);
+
+    $report = (new SmsDeliveryReport())
+      ->setMessageId($json['messageId'])
+      ->setTimeDelivered($time->format('U'));
+
+    $result = (new SmsMessageResult())
+      ->setReports([$report]);
+
+    $message = (new SmsMessage())
+      ->setMessage($json['content'])
+      ->setDirection(Direction::INCOMING)
+      ->setGateway($sms_gateway)
+      ->setResult($result)
+      ->setOption('telstra_status', $json['status']);
+
+    $response = new Response('', 204);
+    $task = (new SmsProcessingResponse())
+      ->setResponse($response)
+      ->setMessages([$message]);
+
+    return $task;
+  }
+
+  /**
+   * Authenticate.
+   *
+   * @return string|FALSE
+   *   Access token for use in auth requests or FALSE if failed.
+   *
+   * @throws RequestException
+   *   Throws on failed HTTP request.
+   */
+  private function telstra_authenticate() {
+    $client_id = $this->configuration['consumer_key'];
+    $secret = $this->configuration['consumer_secret'];
+
+    $result = $this->authenticationApi->authToken($client_id, $secret, 'client_credentials');
+
+    return $result->getAccessToken();
+  }
+
+  /**
+   * Provision a number.
+   *
+   * @return string|FALSE
+   *   Phone number provisioned to app.
+   *
+   * @throws RequestException
+   *   Throws on failed HTTP request.
+   */
+  private function telstra_provision() {
+    $this->provisioningApi->getConfig()->setAccessToken($this->telstra_authenticate());    
+    $provision_request = new \Telstra_Messaging\Model\ProvisionNumberRequest();
+
+    $provision_response = $this->provisioningApi->createSubscription($provision_request);
```

Comment:
Use getSubscription() to check for existing subscription before creating a new one.

Reply:
That said, calling createSubscription() again can extend the life of a number.